core/embed/lib/gfx_draw.c

```diff
@@ -241,6 +241,53 @@ void gfx_draw_text(gfx_offset_t pos, const char* text, size_t maxlen,
   }
 }
 
+#ifdef TREZOR_PRODTEST
+
+#include "qr-code-generator/qrcodegen.h"
+#define QR_MAX_VERSION 9
+
+void gfx_draw_qrcode(int x, int y, uint8_t scale, const char* data) {
+  if (scale < 1 || scale > 10) return;
+
+  uint8_t codedata[qrcodegen_BUFFER_LEN_FOR_VERSION(QR_MAX_VERSION)] = {0};
+  uint8_t tempdata[qrcodegen_BUFFER_LEN_FOR_VERSION(QR_MAX_VERSION)] = {0};
+
+  int side = 0;
+  if (qrcodegen_encodeText(data, tempdata, codedata, qrcodegen_Ecc_MEDIUM,
+                           qrcodegen_VERSION_MIN, QR_MAX_VERSION,
+                           qrcodegen_Mask_AUTO, true)) {
+    side = qrcodegen_getSize(codedata);
+  }
+
+  // Calculate border size (1 extra modules around the QR code)
+  int border_side = ((side + 2) * scale);
+
+  // Calculate border left-top corner
+  x -= border_side / 2;
+  y -= border_side / 2;
+
+  // Draw border around the QR code
```

Comment:
IIUC this doesn't just draw the border but fills the entire area with white color, which is only then filled with black where at appropriate spots. So the comment is misleading.

Reply:
fixed here c1d9151b17c3b20b9f9b508148e44f8f98503eb9